constraint/pkg/client/drivers/k8scel/driver_test.go

```diff
@@ -37,38 +37,29 @@ func makeTemplateWithSource(source *schema.Source, vapGenerationVal *string) *te
 			},
 		},
 	}
-	if vapGenerationVal != nil {
-		template.SetLabels(map[string]string{
-			VAPGenerationLabel: *vapGenerationVal,
-		})
-	}
 	return template
 }
 
-func makeTemplate(vapGenerationVal *string) *templates.ConstraintTemplate {
+func makeTemplate(vapGenerationVal *bool) *templates.ConstraintTemplate {
 	return makeTemplateWithSource(&schema.Source{
 		Validations: []schema.Validation{
 			{
 				Expression: "1 == 1",
 				Message:    "Always true",
 			},
 		},
-	}, vapGenerationVal)
+		GenerateVAP: vapGenerationVal,
+	})
 }
 
-func makeConstraint(vapGenerationVal *string) *unstructured.Unstructured {
+func makeConstraint() *unstructured.Unstructured {
```

Comment:
I forgot how can users override at the constraint level? @JaydipGabani  can you pls add this here and to the design doc?

Reply:
@ritazh we decided on inheriting behavior from template for constraints. If template has `generateVAP` set to true, we generate VAPB for constraints. And since we are treating VAP as enforcement point we would have it enabled by default and users can ommit VAP with `scopedEnforcementActions` spec field on constraints. But generation property is now inherited from template.

Reply:
My bad, should have been clear. With scopedEnforcementActions, user can decide which engine to use, CEL or Rego, for enforcing policies. However generation property is always controlled by `generateVAP` field. Once VAP and VAPB exists in the cluster, intree enforcement will always be there. 

Giving the design a 2nd thought I think this would be a better approach - 

1. `generateVAP` is set to true in CT, generate VAP.
2. `generateVAP` is set to true in CT, generate VAPB for C.
3. `generateVAP` is set to true in CT, VAP enforcementPoint is ommited

Reply:
> With scopedEnforcementActions, user can decide which engine to use, CEL or Rego

nit: to be clear, ^ this statement is incorrect. `scopedEnforcementActions` helps users define for a given constraint, which enforcementPoint (gatekeeper webhook or vap/intree or gator) and which action (deny, warn) can be used. users choose which engine to with the following in the constraint template:

```
targets:
    - target: admission.k8s.gatekeeper.sh
      code:
```

Reply:
Also, bindings would not be generated/not generated by CT.

The logic would be something like:

- If `generateVAP` is set to true in the CT, generate VAP
- If VAP enforcement point is omitted on a constraint, do not generate a VAP Binding
- If VAP enforcement point is enabled on a constraint, generate a VAP binding ONLY if `generateVAP` is true on the CT (arguably we could generate VAP Bindings regardless and leave it to the user to make sure the template is configured appropriately, this may be good for failing closed on a misconfiguration -- thoughts?)

Where a constraint is evaluated is gated purely by the enforcement point. We should not try to infer. Simply: if the user says enforce via Gatekeeper, do that. If the user says enforce via VAP, do that.

In general, we should not try to paper over misconfigurations by trying to infer intent, doing so leads to spaghetti code and weird corner cases.

I realize this is a different approach than before (particularly if we ALWAYS generate a binding regardless of the contents of the template), but multi-enforcement-point lets us avoid the hacky "try to infer things through annotations" we were forced to use before.

Reply:
I always thought it’s odd that k8s allows creations of VAPB without VAP. do you recall why? If we allow VAPB to be created regardless of CT misconfiguration, we miss an opportunity to help users to troubleshoot.  If VAPB got created correctly then it might be harder for the operator to know something is wrong with the CT.

Reply:
It may be a principled stance against imperative-type configuration?

>  If we allow VAPB to be created regardless of CT misconfiguration, we miss an opportunity to help users to troubleshoot.

Not necessarily. In fact, with the old behavior we were actually masking potentially undesired state... we used whichever enforcement point made sense given the combination of annotations. A user had no way of saying "use VAP or bust", which could have led to unintentional webhook usage.

Without running inference, we are giving users a bit more power by affirmatively saying "I want VAP". It does come at the cost of now the user needs to be aware of whether VAP is enabled, but Gatekeeper couldn't tell if a VAP definition was successfully installed anyway, since they don't have `status` (it may be that there were compilation errors). The user will also have more control over the correct approach for "enforcement gaps" (as defined by the [user flow doc](https://docs.google.com/document/d/1UELB0dmt10K5jBwETUnSKKQN7qeivOjvi-AvzFTOsHw/edit#heading=h.qr0os9voygyv).

Without the inference logic, what G8r will do is much more simply explained, which should be both easier to write and easier to debug. Eventually users will even be able to use a "fail"-type failure policy to make misconfigurations even more visible.

WRT simplifying debugging...

we can always report a "missing definition" state on the status of the constraint.

We could think about something like webhook validation, but that also has faults. It will lead to transient rejections if users kubectl apply a bulk manifest. It also wouldn't address situations where the definition is removed out from under the constraint.

Perhaps the biggest reason for separating the two is coherence. Either webhook and VAP are separate enforcement points or they are not. I think we already had this discussion and said that they were separate? There was an option for a three-EP approach where we had VAP, VWH and "whichever works", I don't think we opted for that last one, though.

Reply:
I'm ok with always use enforcementAction to determine VAPB generation or not, instead of deciding for users with what is provided in the constraint template. 

Default enforcementAction would include VAP therefore always generate VAPB. If user explicitly omitted VAP from enforcementPoint in constraint, then do not generate VAPB.